iup/all_objects.py

```diff
@@ -0,0 +1,366 @@
+            {"region", "date", "interval", "estimate"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert (
+            self["region"].dtype == pl.Utf8
+        ), "Column 'region' and/or 'season' is not a string."
+        assert all(
+            self[x].dtype.is_numeric() for x in ["interval", "estimate"]
+        ), "Column 'elapsed', 'interval', 'estimate' and/or 'previous' is not numeric."
+
+
+class CumulativeUptakeData(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "estimate"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert self["region"].dtype == pl.Utf8, "Column 'region' is not a string."
+        assert self["estimate"].dtype.is_numeric(), "Column 'estimate' is not numeric."
+        assert (
+            self["estimate"] >= 0.0
+        ).all(), "Cumulative uptake estimates cannot be negative."
+
+    def to_incident(self, rollout: dt.datetime) -> IncidentUptakeData:
+        self = (
+            self.with_columns(
+                season=pl.col("date").dt.year()
+                + pl.when(pl.col("date").dt.month() < 7).then(-1).otherwise(0),
+                estimate=pl.col("estimate")
+                .diff()
+                .fill_null(pl.col("estimate").first()),
+                elapsed=(pl.col("date") - rollout).cast(pl.Float64)
+                / (1000 * 60 * 60 * 24),
+            )
+            .with_columns(
+                interval=pl.col("elapsed").diff().fill_null(pl.col("elapsed").first()),
+                season=pl.col("season").cast(pl.Utf8),
+            )
+            .with_columns(estimate=pl.col("estimate") / pl.col("interval"))
+            .with_columns(previous=pl.col("estimate").shift(1))
+        )
+
+        if ((self["interval"] - self["interval"].mean()) / self["interval"].std())[
+            0
+        ] > 1:
+            self = self.slice(2, self.height - 1)
+        else:
+            self = self.slice(1, self.height - 1)
+
+        self = IncidentUptakeData(self)
+        self.validate()
+
+        return self
+
+
+class CumulativeUptakeProjection(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "estimate"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert self["region"].dtype == pl.Utf8, "Column 'region' is not a string."
+        assert self["estimate"].dtype.is_numeric(), "Column 'estimate' is not numeric."
+
+
+def get_nis(
+    data_path, region_col, date_col, estimate_col, filters=None
+) -> CumulativeUptakeData:
+    out = (
+        (pl.read_csv(data_path))
+        .with_columns(estimate=pl.col(estimate_col).cast(pl.Float64, strict=False))
```

Comment:
There's usually a bunch of other columns that get dropped anyway, so it's a matter of `.select()`ing the few columns I want, rather than dropping all the ones I don't. Still, it's good to learn an alternative method!